server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java

```diff
@@ -0,0 +1,130 @@
+
+    private static final Logger log = LoggerFactory.getLogger(NanoExtension.class);
+
+    public static final CryptoCurrency CRYPTO = CryptoCurrency.NANO;
+
+    private volatile NanoExtensionContext context = new NanoExtensionContext(CRYPTO, ctx, NanoUtil.NANO);
+
+
+    @Override
+    public String getName() {
+        return "BATM Nano extra extension";
+    }
+
+    @Override
+    public void init(IExtensionContext ctx) {
+        super.init(ctx);
+        this.context = new NanoExtensionContext(CRYPTO, ctx, NanoUtil.NANO);
+    }
+
+    @Override
+    public IWallet createWallet(String walletLogin, String tunnelPassword) {
+        try {
+            if (walletLogin != null && !walletLogin.trim().isEmpty()) {
+                StringTokenizer st = new StringTokenizer(walletLogin, ":");
+                String walletName = st.nextToken();
+
+                if ("nano_node".equalsIgnoreCase(walletName)) {
+                    return NanoNodeWallet.create(context, st);
+                } else if ("nano_demo".equalsIgnoreCase(walletName)) {
+                    String fiatCurrency = st.nextToken();
+                    String walletAddress = st.nextToken();
+                    return new DemoWallet(fiatCurrency, CRYPTO.getCode(), walletAddress);
+                }
+            }
+        } catch (Exception e) {
+            log.error("Couldn't create wallet.", e);
+        }
+        return null;
+    }
+
+    @Override
+    public ICryptoAddressValidator createAddressValidator(String cryptoCurrency) {
+        if (CRYPTO.getCode().equalsIgnoreCase(cryptoCurrency))
+            return new NanoAddressValidator(context);
+        return null;
+    }
+
+    @Override
+    public IRateSource createRateSource(String sourceLogin) {
+        if (sourceLogin != null && !sourceLogin.trim().isEmpty()) {
+            StringTokenizer st = new StringTokenizer(sourceLogin, ":");
+            String sourceType = st.nextToken();
+
+            if ("coinmarketcap".equalsIgnoreCase(sourceType)) {
+                String preferredCurrency = getPreferredCurrency(st, FiatCurrency.USD);
+                String apiKey = st.hasMoreTokens() ? st.nextToken() : null;
+                return new CoinmarketcapRateSource(apiKey, preferredCurrency);
+            } else if ("coingecko".equalsIgnoreCase(sourceType)) {
+                return new CoinGeckoRateSource(getPreferredCurrency(st, FiatCurrency.USD));
+            } else if ("coinpaprika".equalsIgnoreCase(sourceType)) {
+                return new CoinPaprikaRateSource(getPreferredCurrency(st, FiatCurrency.USD));
+            } else if ("binancecom".equalsIgnoreCase(sourceType)) {
+                return new BinanceComExchange(getPreferredCurrency(st, FiatCurrency.EUR));
+            } else if ("binanceus".equalsIgnoreCase(sourceType)) {
+                return new BinanceUsExchange(getPreferredCurrency(st, FiatCurrency.USD));
+            }
+        }
+        return null;
+    }
```

Comment:
Please don't add the rate sources that are already created in other extensions. 
If you add NANO to coinmarketcap rate source for example and `new CoinmarketcapRateSource` is called in another extension it will work fine.

Reply:
Wouldn't that mean our extension is coupled to the Dash implementation for instance? So for example, if `DashExtension` specifies `CoinmarketcapRateSource` as of now, and removes it at a later date, that will affect the behaviour of the `NanoExtension` (and any other extension that relies on that source). Almost all extensions seem to re-specify the rate sources (at least 15 extensions).

That being said, I'll remove it for the time being if that's what you'd prefer.

Reply:
Thanks. You're right coinmarketcap is re-specified where it's not necessary in other extensions. But we will not remove it in future, all the extensions in this repository are distributed together as part of the "server_extensions_extra" extensions JAR file.